uxarray/grid/validation.py

```diff
@@ -65,3 +67,11 @@ def _check_area(self):
     else:
         print("-No face area is close to zero.")
         return True
+
+
+@njit(cache=True)
+def _mesh_contains_holes(edge_face_connectivity):
+    """Check if a mesh has holes in it."""
+
+    # If an edge only has one face saddling it than the mesh has holes in it
+    return np.any(edge_face_connectivity[:, 1] == INT_FILL_VALUE)
```

Comment:
I don't see an issue with this check, as the reasoning that @aaronzedwick explained is sound. If there exists any edge that does not have two neighboring faces, then there must exist some "empty" area adjacent to that edge. Illustration below for reference: the edges marked in red each neighbor an empty region, which means that we can't construct an edge between the centers of the two faces indicated in the `edge_face_connectivity`

![image](https://github.com/user-attachments/assets/2dc764c8-5080-4992-9f72-977d709cead3)

This causes issues with the Dual Mesh construction, since node-centered data variables become face-centered. However, there will be no faces constructed centered around these values. Given our interpretation of the UGRID conventions, we have no way to construct a face for this data, meaning that the data is essentially "lost" from the dual mesh construction process. 

Also, I will suggest caching the result once we've run it, so that we don't need to run the repeated searches over the `edge_face_connectivity`. Something as simple as a `Grid._partial_coverage` attribute would suffice.

Reply:
> I don't see an issue with this check, as the reasoning that @aaronzedwick explained is sound. If there exists any edge that does not have two neighboring faces, then there must exist some "empty" area adjacent to that edge. Illustration below for reference: the edges marked in red each neighbor an empty region, which means that we can't construct an edge between the centers of the two faces indicated in the `edge_face_connectivity`

We have already gone through this part in the previous discussion above. I understand what you mean by "missing a face" here. My current concern is: after we have examined the entire grid to check if there's a partial coverage, what's the next step to deal with the missing faces? Do we need to rerun a similar check to identify which faces don't have neighbors? So basically two O(N) operations?


>Also, I will suggest caching the result once we've run it, so that we don't need to run the repeated searches over the `edge_face_connectivity`. Something as simple as a `Grid._partial_coverage ` attribute would suffice.

This is a good point. As we agreed in the long conversation from the `zonal_mean` PR, we need to provide the sudo codes for the algorithm. So that we can better adapt it to the UXarray API and the python vectorization styles.

Reply:
> My current concern is: after we have examined the entire grid to check if there's a partial coverage, what's the next step to deal with the missing faces? Do we need to rerun a similar check to identify which faces don't have neighbors? So basically two O(N) operations?

This is a good point that I overlooked. We should be keeping track of which indices are considered to border an empty area, which can be achieved by removing the `np.any` call
```Python
edge_face_connectivity[:, 1] == INT_FILL_VALUE
```

Once we have this, we need to identify which nodes (indicated in black) will need to be removed from the new grid that is constructure from the Dual Grid. 

![image](https://github.com/user-attachments/assets/6c82d022-38ef-445d-b2ac-28a49618b16c)

The number of faces in the new `Grid` should equal the original number of nodes (`n_node`) minus the number of nodes that share one or more of the "invalid" edges shown above.

Reply:
```python
edge_face_connectivity[:, 1] == INT_FILL_VALUE
```

This sounds good. If we need to check if there's a partial coverage, don't just return a boolean value, but instead directly return where is the partial coverage. And if the return is `None`, then we know is all covered.


> The number of faces in the new Grid should equal the original number of nodes (n_node) minus the number of nodes that share one or more of the "invalid" edges shown above.

It sounds good to me though I haven't thought deep into this yet. But I will keep you updated if anything other things comes into my head.

Reply:
Sounds perfect to me, @philipc2 @hongyuchen1030 thanks for the suggestions on this one! Let me know if you have anymore thoughts and I will get the changes mentioned here implemented!